File: java-pine/src/main/java/page/liam/pine/operators/TransformByLua.java
package page.liam.pine.operators;

import page.liam.pine.AbstractOperator;
import page.liam.pine.OperatorInput;
import page.liam.pine.OperatorOutput;
import org.luaj.vm2.*;
import org.luaj.vm2.lib.jse.JsePlatform;

import java.util.Map;

public class TransformByLua extends AbstractOperator {
    private String script;
    private String funcName;
    private boolean isItemMode;

    @Override
    public void init(Map<String, Object> params) throws Exception {
        script = (String) params.get("lua_script");
        String funcForItem = (String) params.getOrDefault("function_for_item", "");
        String funcForCommon = (String) params.getOrDefault("function_for_common", "");

        if (funcForItem.isEmpty() && funcForCommon.isEmpty()) {
            throw new IllegalArgumentException("lua: exactly one of function_for_item or function_for_common must be set");
        }
        if (!funcForItem.isEmpty() && !funcForCommon.isEmpty()) {
            throw new IllegalArgumentException("lua: cannot set both function_for_item and function_for_common");
        }

        if (!funcForItem.isEmpty()) {
            funcName = funcForItem;
            isItemMode = true;
        } else {
            funcName = funcForCommon;
            isItemMode = false;
        }

        // Validate script compiles
        Globals globals = JsePlatform.standardGlobals();
        globals.load(script).call();
    }

    @Override
    public void execute(OperatorInput input, OperatorOutput output) throws Exception {
        Globals globals = JsePlatform.standardGlobals();
        globals.load(script).call();

        if (isItemMode) {
            executeForItem(globals, input, output);
        } else {
            executeForCommon(globals, input, output);
        }
    }

    private void executeForItem(Globals globals, OperatorInput input, OperatorOutput output) throws Exception {
        // Set common globals once
        for (String field : commonInput) {
            globals.set(field, toLua(input.common(field)));
        }

        LuaValue fn = globals.get(funcName);
        if (fn.isnil()) {
            throw new Exception("lua: function \"" + funcName + "\" not found");
        }

        int nret = itemOutput.size();
        int n = input.itemCount();

        for (int i = 0; i < n; i++) {
            // Set item globals for this item
            for (String field : itemInput) {
                globals.set(field, toLua(input.item(i, field)));
            }

            Varargs results = fn.invoke(LuaValue.NONE);

            for (int j = 0; j < nret; j++) {
                Object val = toJava(results.arg(j + 1));
                output.setItem(i, itemOutput.get(j), val);
            }
        }
    }

    private void executeForCommon(Globals globals, OperatorInput input, OperatorOutput output) throws Exception {
        // Set common globals as scalars
        for (String field : commonInput) {
            globals.set(field, toLua(input.common(field)));
        }

        // Set item fields as 1-indexed Lua tables
        int n = input.itemCount();
        for (String field : itemInput) {
            LuaTable tbl = new LuaTable();
            for (int i = 0; i < n; i++) {
                tbl.set(i + 1, toLua(input.item(i, field)));
            }
            globals.set(field, tbl);
        }

        LuaValue fn = globals.get(funcName);
        if (fn.isnil()) {
            throw new Exception("lua: function \"" + funcName + "\" not found");
        }

        int nret = commonOutput.size();
        Varargs results = fn.invoke(LuaValue.NONE);

        for (int j = 0; j < nret; j++) {
            Object val = toJava(results.arg(j + 1));
            output.setCommon(commonOutput.get(j), val);
        }
    }

    private static LuaValue toLua(Object v) {
        if (v == null) return LuaValue.NIL;
        if (v instanceof Boolean) return LuaValue.valueOf((Boolean) v);
        if (v instanceof Number) return LuaValue.valueOf(((Number) v).doubleValue());
        if (v instanceof String) return LuaValue.valueOf((String) v);
        return LuaValue.valueOf(String.valueOf(v));
    }

    private static Object toJava(LuaValue v) {
        if (v.isnil()) return null;
        if (v.isboolean()) return v.toboolean();
        if (v.isnumber()) return v.todouble();
        if (v.isstring()) return v.tojstring();
        return v.tojstring();
    }
}

File: pine-cpp/operators/transform/transform_by_lua.cpp
#include "pine/operator.hpp"

#include "lua/lua_bridge.hpp"
#include "lua/lua_pool.hpp"
#include "operators/_helpers.hpp"

namespace pine {

class TransformByLuaOp : public Operator, public ConcurrentSafe, public StatsProvider, public MetricsAware {
 public:
  void init(const OperatorConfig& cfg) override {
    op_name_ = cfg.name;
    const auto& params = cfg.params.as_object();
    auto script_it = params.find("lua_script");
    if (script_it == params.end() || !script_it->second.is_string()) {
      throw RegistryError(op_name_,
                          "lua: exactly one of function_for_item or function_for_common must be set");
    }
    lua_script_ = script_it->second.as_string();

    auto fi_it = params.find("function_for_item");
    auto fc_it = params.find("function_for_common");
    func_for_item_ = (fi_it != params.end() && fi_it->second.is_string()) ? fi_it->second.as_string() : "";
    func_for_common_ = (fc_it != params.end() && fc_it->second.is_string()) ? fc_it->second.as_string() : "";

    if (func_for_item_.empty() && func_for_common_.empty()) {
      throw RegistryError(op_name_,
                          "lua: exactly one of function_for_item or function_for_common must be set");
    }
    if (!func_for_item_.empty() && !func_for_common_.empty()) {
      throw RegistryError(op_name_, "lua: cannot set both function_for_item and function_for_common");
    }

    common_input_ = cfg.metadata.common_input;
    item_input_ = cfg.metadata.item_input;
    item_output_ = cfg.metadata.item_output;
    common_output_ = cfg.metadata.common_output;

    pool_ = std::make_unique<lua::StatePool>(lua_script_, op_name_);
  }

  void set_metrics_provider(metrics::Provider* provider) override {
    if (pool_) {
      pool_->set_metrics(provider, op_name_);
    }
  }

  std::map<std::string, int64_t> operator_stats() const override {
    if (pool_) {
      return pool_->stats_snapshot();
    }
    return {};
  }

  void execute(const OperatorInput& input, OperatorOutput& out) override {
    auto borrowed = pool_->borrow();
    lua::LuaVM& vm = *borrowed;

    if (!func_for_item_.empty()) {
      int nret = static_cast<int>(item_output_.size());
      for (const auto& field : common_input_) {
        vm.set_global(field, input.common(field));
      }
      for (std::size_t i = 0; i < input.item_count(); ++i) {
        for (const auto& field : item_input_) {
          vm.set_global(field, input.item(i, field));
        }
        auto results = vm.call_function(func_for_item_, nret, op_name_);
        for (int j = 0; j < nret; ++j) {
          out.set_item(static_cast<int>(i), item_output_[static_cast<std::size_t>(j)],
                       results[static_cast<std::size_t>(j)]);
        }
      }
    } else {
      int nret = static_cast<int>(common_output_.size());
      for (const auto& field : common_input_) {
        vm.set_global(field, input.common(field));
      }
      for (const auto& field : item_input_) {
        std::vector<JsonValue> column;
        column.reserve(input.item_count());
        for (std::size_t i = 0; i < input.item_count(); ++i) {
          column.push_back(input.item(i, field));
        }
        vm.set_global_table(field, column);
      }
      auto results = vm.call_function(func_for_common_, nret, op_name_);
      for (int j = 0; j < nret; ++j) {
        out.set_common(common_output_[static_cast<std::size_t>(j)], results[static_cast<std::size_t>(j)]);
      }
    }
  }

 private:
  std::string op_name_;
  std::string lua_script_;
  std::string func_for_item_;
  std::string func_for_common_;
  std::vector<std::string> common_input_;
  std::vector<std::string> item_input_;
  std::vector<std::string> item_output_;
  std::vector<std::string> common_output_;
  std::unique_ptr<lua::StatePool> pool_;
};

static const OperatorSchema k_transform_by_lua_schema{
    .name = "transform_by_lua",
    .type = OpType::Transform,
    .description = "Executes a Lua script for per-item or per-common computation.",
    .params =
        {
            {"function_for_common",
             {.type = "string",
              .required = false,
              .default_value = JsonValue(""),
              .description = "Function name to call once for all items."}},
            {"function_for_item",
             {.type = "string",
              .required = false,
              .default_value = JsonValue(""),
              .description = "Function name to call per item."}},
            {"lua_script",
             {.type = "string",
              .required = true,
              .default_value = JsonValue(nullptr),
              .description = "Lua source code defining the function to call."}},
        },
};
PINE_REGISTER_OPERATOR_T(TransformByLuaOp, k_transform_by_lua_schema)

}  // namespace pine
